core/src/banking_stage/latest_unprocessed_votes.rs

```diff
@@ -0,0 +1,1036 @@
+
+        let instruction_filter = |ix: &VoteInstruction| {
+            if deprecate_legacy_vote_ixs {
+                matches!(
+                    ix,
+                    VoteInstruction::TowerSync(_) | VoteInstruction::TowerSyncSwitch(_, _),
+                )
+            } else {
+                ix.is_single_vote_state_update()
+            }
+        };
+
+        match limited_deserialize::<VoteInstruction>(&instruction.data) {
+            Ok(vote_state_update_instruction)
+                if instruction_filter(&vote_state_update_instruction) =>
+            {
+                let vote_account_index = instruction
+                    .accounts
+                    .first()
+                    .copied()
+                    .ok_or(DeserializedPacketError::VoteTransactionError)?;
+                let vote_pubkey = message
+                    .message
+                    .static_account_keys()
+                    .get(vote_account_index as usize)
+                    .copied()
+                    .ok_or(DeserializedPacketError::VoteTransactionError)?;
+                let slot = vote_state_update_instruction.last_voted_slot().unwrap_or(0);
+                let hash = vote_state_update_instruction.hash();
+                let timestamp = vote_state_update_instruction.timestamp();
+
+                Ok(Self {
+                    vote: Some(vote),
+                    slot,
+                    hash,
+                    vote_pubkey,
+                    vote_source,
+                    timestamp,
+                })
+            }
+            _ => Err(DeserializedPacketError::VoteTransactionError),
+        }
+    }
+
+    pub fn get_vote_packet(&self) -> Arc<ImmutableDeserializedPacket> {
+        self.vote.as_ref().unwrap().clone()
+    }
+
+    pub fn vote_pubkey(&self) -> Pubkey {
+        self.vote_pubkey
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.slot
+    }
+
+    pub(crate) fn hash(&self) -> Hash {
+        self.hash
+    }
+
+    pub fn timestamp(&self) -> Option<UnixTimestamp> {
+        self.timestamp
+    }
+
+    pub fn is_vote_taken(&self) -> bool {
+        self.vote.is_none()
+    }
+
+    pub fn take_vote(&mut self) -> Option<Arc<ImmutableDeserializedPacket>> {
+        self.vote.take()
+    }
+}
+
+#[derive(Default, Debug)]
+pub(crate) struct VoteBatchInsertionMetrics {
+    pub(crate) num_dropped_gossip: usize,
+    pub(crate) num_dropped_tpu: usize,
+}
+
+impl VoteBatchInsertionMetrics {
+    pub fn total_dropped_packets(&self) -> usize {
+        self.num_dropped_gossip + self.num_dropped_tpu
+    }
+
+    pub fn dropped_gossip_packets(&self) -> usize {
+        self.num_dropped_gossip
+    }
+
+    pub fn dropped_tpu_packets(&self) -> usize {
+        self.num_dropped_tpu
+    }
+}
+
+#[derive(Debug)]
+pub struct LatestUnprocessedVotes {
+    latest_vote_per_vote_pubkey: RwLock<HashMap<Pubkey, Arc<RwLock<LatestValidatorVotePacket>>>>,
```

Comment:
You can use `PubkeyHasherBuilder` to make this HashMap 4x faster, its a 1 line change (see https://github.com/anza-xyz/solana-sdk/pull/96 for context)

Reply:
Fair point.